### ISSUE TITLE
binaries failing to exec on macOS with prog mmap errno 1

### ISSUE THREAD

**Issue by paulsmith** [jart/cosmopolitan#918]
I downloaded the latest cosmos-3.0.1 release on my M1 MacBookPro running Sonoma, unzipped, cd'd to bin, and ran vim - hopped over to System Settings -> Privacy & Security -> clicked "Allow Anyway", re-ran vim, approved the "cannot verify developer" pop-up dialog, and got the following

```
$ ./vim
ape error: ./vim: prog mmap failed w/ errno 1
```

This is reproducible with any of the binaries in the bin directory.

System details:

- Darwin 23.1.0 arm64
- MacBookPro18,3

**Comment by DylanLukes** [comment]
Can also report the same thing with `make.com` (meaning I can't build the toolchain).

```
/opt/cosmo master ❯ ./build/bootstrap/make.com
ape error: ./build/bootstrap/make.com: prog mmap failed w/ errno 1
```

System details:

  - Darwin 23.0.0 aarch64
  - MacBookPro18,2

**Comment by DylanLukes** [comment]
What follows is a summary of findings discussed on Discord.

**tl;dr:** this is due to SIP. If you [disable SIP](https://developer.apple.com/documentation/security/disabling_and_enabling_system_integrity_protection), everything works. Things look optimistic that this _may_ not always be required though!

### The SIP/`mmap` weirdness

On macOS 14 Sonoma, it appears that a component involved with enforcing SIP, AppleMobileFileIntegrity (AMFI), has some strange behavior when it comes to allowing `mmap` with `PROT_EXEC`.

Namely: _`mmap` will fail with EPERM (-1) for any executable which has not previously been run at least once, or any shared library which has not previously had dlopen called on it._

In the following, `mmap-test` just attempts to `PROT_READ | PROT_EXEC` the contents of `dummy`, which is just the standard C hello world.  Seeing is believing... 

```sh
❯ cc -o mmap-test mmap-test.c               
❯ cc -o dummy dummy.c
❯ ./mmap-test
size: 33432
mmap (simple) failed w/ 1
[1]    31056 segmentation fault  ./mmap-test
❯ ./dummy
Hello, world!
❯ ./mmap-test
size: 33432
```

You can find the full reproduction of that here: [DylanLukes/mmap-weirdness](https://github.com/DylanLukes/mmap-weirdness).

This does _not_ appear to have anything to do with the code signature on the binary or dynamic library itself, but rather some SIP related internal book-keeping. **This behavior vanishes when SIP is disabled (and cosmo will work too!)**

The system logs corroborate this. We can see that for the first `mmap-test` run:

```
error    23:44:30.439147-0700    kernel    Library Validation failed: Rejecting '/Users/dylan/Projects/mmap-weirdness/dummy' (Team ID: none, platform: no) for process 'mmap-test(4513)' (Team ID: N/A, platform: no), reason: Trying to load an unsigned library
```

Running `dummy` the first time produces these logs:

```
default    23:44:36.289709-0700    kernel    AMFI: '/Users/dylan/Projects/mmap-weirdness/dummy' has no CMS blob?
default    23:44:36.289716-0700    kernel    AMFI: '/Users/dylan/Projects/mmap-weirdness/dummy': Unrecoverable CT signature issue, bailing out.
```

However, after this first run, these messages will not reoccur if you re-run `dummy` (or `mmap-test`).

### Relevance to APE loading

The issue here for Cosmo is that we have a chicken-and-egg problem: we can't "run" an APE (`.com`) first to get the system to mark it as "safe to `PROT_EXEC` `mmap`".  In order to do that, we'd need it to be runnable, which it isn't because the `ape-m1.c` loader can't run a `.com` without performing this `mmap`.

### Bypassing (but not for Cosmo)

This change in behavior can be triggered either by:

1. Running it, if it's an executable.
2. Calling `dlopen`, if it's a `.dylib`.

Amusingly, the `dlopen` will _also_ work even if it's an executable!

So, armed with the following `dlhack.c`...

```c
#include <dlfcn.h>
int main(int argc, char **argv) { dlopen(argv[1], RTLD_LAZY); }
```

...we can get macOS to let us `mmap` with `PROT_EXEC` an  arbitrary executable or shared library.

```sh
❯ cc -o dummy dummy.c
❯ ./mmap-test dummy
size: 33432
mmap (simple) failed w/ 1
❯ ./dlhack dummy
❯ ./mmap-test dummy
size: 33432
mmap (simple) succeeded
```

### What about for cosmo...?

The above trick is probably not going to work. However, as noted by @jart, it's probably possible to bypass this with `MAP_JIT` instead.

That said, it's fun to consider that a potentially viable (but likely limited) way to bypass this would be:

1. Write the relevant executable portion of the `.com` to a temporary file.
2. Call `dlopen` on that temporary file.
3. Call `mmap` on temporary file with `PROT_EXEC`.
4. Proceed.

Now that's what I call _security_.

**Comment by DylanLukes** [comment]
This has been fixed in a12ad17! 🎉

**Comment by pkulchenko** [comment]
Great; thank you for the analysis and the confirmation!

**Comment by kassiansun** [comment]
Hi @DylanLukes which version of cosmocc should I download? All zips on https://cosmo.zip/pub/cosmocc/ is not working:
- [cosmocc-0.0.18.zip](https://cosmo.zip/pub/cosmocc/cosmocc-0.0.18.zip) is used as part of cosmo but I can't get it running at the moment
- The other two packages all have `prog mmap failed w/ errno 1` when running `cosmocc`.

I didn't disable SIP, is it a still required for the latest cosmocc build?

**Comment by jart** [comment]
I'm running https://cosmo.zip/pub/cosmocc/cosmocc.zip on a system with SIP enabled just fine this past week. I haven't bumped the version number yet of the `ape` command, so you may need to `rm -rf $TMPDIR/.ape*` before running the new binaries. It's recommended that you run the `ape/apeinstall.sh` script in the cosmo repository. You could also just `cc cosmocc/bin/ape-m1.c -o /usr/local/bin/ape`.

**Comment by kassiansun** [comment]
Tried the latest zip and followed these instructions, now it works fine, thanks!
